src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -569,11 +605,73 @@ FlexCPUThread::markFault(shared_ptr<InflightInst> inst_ptr, Fault fault)
     commitAllAvailableInstructions();
 }
 
+void
+FlexCPUThread::onBranchPredictorAccessed(std::weak_ptr<InflightInst> inst,
+                                         BPredUnit* pred)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr
+     || inst_ptr->isCommitted()
+     || inst_ptr->isSquashed()
+     || inst_ptr->isComplete()) {
+        freeBranchPredDepth();
+        // No need to do anything for an instruction that has been
+        // squashed, or a branch that has already been resolved.
+        return;
+    }
+
+    // We check before requesting one from the CPU, so this function should
+    // only be called with a non-NULL branch predictor.
+    panic_if(!pred, "I wasn't programmed to understand nullptr branch "
+                    "predictors!");
+
+    TheISA::PCState pc = inst_ptr->pcState();
+
+    const bool taken = pred->predict(inst_ptr->staticInst(),
+                                     inst_ptr->issueSeqNum(), pc, threadId());
+
+    // BPredUnit::predict takes pc by reference, and updates it in-place.
+
+    DPRINTF(FlexCPUBranchPred, "(seq %d) predicted %s (predicted pc: %#x).\n",
+                              inst_ptr->seqNum(),
+                              taken ? "taken" : "not taken",
+                              pc.instAddr());
+    // Count stats using the predict return value?
+
+    StaticInstPtr static_inst = inst_ptr->staticInst();
+    if (static_inst->isDirectCtrl() && static_inst->isUncondCtrl()) {
+        // If we *know* we've mispredicted, no need to keep fetching
+        // new instructions. This optimization improves the simulator's
+        // performance, but may underestimate the instruction fetch bandwidth
+        if (!taken || static_inst->branchTarget(inst_ptr->pcState()) != pc) {
+            DPRINTF(FlexCPUBranchPred, "Unconditional branch mispredicted\n");
+            freeBranchPredDepth();
+            return;
+        }
+    } else if ((static_inst->isCall() || static_inst->isReturn()) && !taken) {
+        // Calls are always taken.
+        DPRINTF(FlexCPUBranchPred, "Call/return taken misprediction\n");
+        freeBranchPredDepth();
+        return;
+    }
+
+    inst_ptr->addSquashCallback([this, inst] {
+        const shared_ptr<InflightInst> inst_ptr = inst.lock();
+        if (!inst_ptr || inst_ptr->isCommitted() || inst_ptr->isComplete())
+            return;
+
+        freeBranchPredDepth();
+    });
+
+
+    advanceInst(pc);
```

Comment:
Does this take into account the fetch bandwidth limit? I think so but...

Reply:
I'm not sure how to address this, since:

1) bandwidth constraints are managed by the CPU, so this still goes through that channel.
2) We don't actually have a configurable fetch bandwidth, but an issue bandwidth,  which this instruction will still have to go through.